optuna/storages/_rdb/storage.py

```diff
@@ -1132,6 +1143,54 @@ def get_all_versions(self) -> List[str]:
 
         return self._version_manager.get_all_versions()
 
+    def record_heartbeat(self, trial_id: int) -> None:
+        with _create_scoped_session(self.scoped_session, True) as session:
+            heartbeat = models.TrialHeartbeatModel.where_trial_id(trial_id, session)
+            if heartbeat is None:
+                heartbeat = models.TrialHeartbeatModel(trial_id=trial_id)
+                session.add(heartbeat)
+            else:
+                heartbeat.heartbeat = session.execute(func.now()).scalar()
```

Comment:
[question]

Is there a risk of deadlock? If `n_jobs` is large, many process attempt to invoke `record_heartbeat`.
In the current implementation, each trial has one record in `trial_heartbeats` and it is repeatedly updated.
It means multiple queries could update records in a short period of time.

https://github.com/optuna/optuna/pull/2190/files#diff-94a2b73a9a2e42da9a12a7f9513e423b3490718f10a527e452a17abfcb73887bR483-R498

Reply:
[Just a note] There is no risk of deadlock since we don't have any lock on `TrialHeatbeatModel` and each trial is processed by one process.